### ISSUE TITLE
Show more info in files list of two commit diff

### ISSUE THREAD

**Issue by mdonatas** [gitextensions/gitextensions#4154]
Think about two situations:
1. You want to merge `master` into your `feature` branch (or vice versa) but you would like to know if others have been working on same files as you did. Maybe to know about possible conflicts in advance or to see if you will need to adapt your changes even if the merge won't conflict.
You select two commits your `feature` `HEAD` and `master` and there are 108 files different and almost all of them from changes introduced in `master`! It's a PITA to find files of interest (the ones you changed) in that list to view their diff.
2. You rebase (but still have old commits accessible (see screenshot for an example)) and experience conflicts. Now you want to know if you didn't screw up. You again select two commits and again - many files have changed, how to easily see the diff you care about (diff of files you changed and not other devs in master).

Tadaaaam (illustrating 2nd case):
![image](https://user-images.githubusercontent.com/483659/32756252-1b76cee4-c8d1-11e7-91b1-e38a35ff08ad.png)

So what this feature does is
1. it finds the `common base` for the two commits
2. picks only files which were changed in both `common base -> 1st commit` and `common base -> 2nd commit` and adds 3 new filtered lists. Basically you can see diffs for those files for what _you_ changed; for what _they_ changed and finally a simple diff between them which is the same as diff in the full list but filtered to common files. I use this third view very often after rebasing or before merge'ing.

If you have questions or want clarifications don't hesitate to ask.
I've been using this feature in my custom builds for ~9 months now and I'm loving it. If there'll be interest I would try to get the code to _production quality_ :)

**Comment by gerhardol** [comment]
Sure interesting. I compare to some base (not necessarily the common ancestor) manually.
How is no base handled?
Is "just" first base or can you select the base? (for example select first-base-second).

(I am doing some cleanups to diff, some are not accepted yet, some are not submitted yet)
Production quality takes time...

**Comment by mdonatas** [comment]
To drive a point home. I had a PR for showing commit info in file history form
![image](https://user-images.githubusercontent.com/483659/37044208-8f0bbc28-216b-11e8-9802-ba9c0a146f73.png)
in the meantime a whole bunch of files got updated during cleanup
![image](https://user-images.githubusercontent.com/483659/37044248-ac16e324-216b-11e8-8227-c00d6658d8d1.png)
(and many more commits such as these)

By selecting my PR commit and latest master I get this:
![image](https://user-images.githubusercontent.com/483659/37044401-114d3ac2-216c-11e8-8530-7983b7f62d19.png)

Where Base to A is what I've changed
Where Base to B is what other have changed
Where Common Files is combined diff

Now this immensely helps me to know in advance how I will need to accommodate changes made in master. This is a merge conflict preview of sorts.

**Comment by mdonatas** [comment]
@gerhardol if there is no common base then logic falls back to regular diff with no filtering. Selecting a base manually could be added too rather easily I suppose. Do you have a scenario in mind where immediate common base is not what you would want?

**Comment by gerhardol** [comment]
I am leaving the details how this feature should work to later. Seem useful.
I see scenarios where the branches differs a lot and may even be separate, but there are a few files that could be shared (in situations where submodules are not a clear fit).

However, please take a look at my existing PRs that will allow multiselecting revisions and see all differences (there is no common diff though, except for merge diffs). The structure in RevisionDiff and FileStatusList has changed some. It would be nice if you reviewed the PRs...

**Comment by gerhardol** [comment]
#4569 is now merged.

**Comment by gerhardol** [comment]
@mdonatas Status?

**Comment by mdonatas** [comment]
I've completed rebasing on master not too long ago after a long pause and still have some quirks to fix. Unfortunately free time is not someone with two kids has a lot of :D
Once the bugs are fixed I'll let you know.

**Comment by RussKie** [comment]
Hear! Hear! 😆

**Comment by gerhardol** [comment]
I took a quick look at this in your branch mdonatas/CustomModificationsNeedToFixDiffsAfterHugeRebase

This breaks select multiselect revision diffs in the current state.

One obvious problem is to describe the involved revisions. Base is not described at all what I see. 

When changing so diff is possible from many to the last selected, the naming was changed to consistently match First and Selected. That was mostly following the Git documentation too from what I could see, some discussions in issues too. You are partly changing to use A/B instead. Clear to me, also in Git documentation but maybe not to the average user.

**Comment by mdonatas** [comment]
>This breaks select multiselect revision diffs in the current state.

I had a suspicion of this but since I haven't ever used extensively a version of GitExtensions with that working I don't even know what it does so I didn't miss it, but it's on my TODO.
Later I'll try to run `master` branch, try selecting 3 revs and see what happens :)

`First` and `Selected`... I'll give an analogy of how it feels to me. Imaging every time you had to breath you'd need to consciously think about expanding and contracting your lungs for every breath as opposed to that just happening. Such is the difference to me between `First`, `Selected` and `A`, `B`. I instinctively perceive `A`, `B` as `was` and `is` or `from` and `to`. But with `First` and `Selected` I need to think and what's worse that's a stress inducing thought because even after thinking I feel like I am making a guess and won't know the outcome. So.. I am fine with this never being changed in GE as I run this branch every day at work and I think I haven't used a public release version of GE for at least 3 years :)

If we moved further with this feature, those commits would surely be extracted from this branch leaving A/B and other stuff in this `"a drawer of everything what I don't like about the official release fixed"` of a branch :)

**Comment by mdonatas** [comment]
Addendum to the `First` and `Selected` rant

![image](https://user-images.githubusercontent.com/483659/45256408-c1a7bb80-b39e-11e8-82f1-2d45cd46be18.png)

**Comment by gerhardol** [comment]
Selected is really short for "Last Selected" and is used in other parts of GE too. If only one revision is selected it is clear.
First is what you compare to, so "before selected". If two revisions are selected, it should be obvious, with one selected it should be clear too.

The names make more sense when only one or two revisions are clicked. If you select more revisions, you are likely more advanced and can figure out what First is - the revisions are described in diff and reset submenues (not Save selected as though). 
Casual users are confused by the diff headers too.

I am not happy with First-Selected either, but believe it is something everyone can understand. On the other hand First-Selected must be translated too, which breaks some intentions.
If you have a good idea how to present the revisions, please suggest a change. I will vote against it if it breaks multicomparision though, but there are other voters too...

**Comment by mdonatas** [comment]
I understand the meanings of all these labels after I think about them and look at rev descriptions. The problem is that this doesn't come intuitively.

My suggestion is to go back to `A` and `B`. Just by looking at the diff, you `get it`™ that `A` is the first revision and `B` is second and these are the changes necessary to go from `A` to `B` aka from `first` to `second`. And the great added bonus is that these labels are
- short
- same length
- perfectly translatable (Greeks could use α and Ω 😁)
- Probably suitable for all situations

**Comment by mdonatas** [comment]
To the point of this proposal I am leaning towards enabling this only when two revs are selected, otherwise you'd need a lot of comparisons between each of selections and that might do more harm than good. Already with two selections there are 3 groups. How does that sound?

**Comment by mdonatas** [comment]
@gerhardol On an unrelated note. Could you give examples of when you use more than 2 selections? I tried (without too much vigor) to find a PR or an Issue where this was implemented but couldn't (sorry).

**Comment by gerhardol** [comment]
> My suggestion is to go back to A and B

Open a separate issue about that. I am divided myself. The naming should be consistent with Git naming and within GE.
The code uses first/selected right now. I started converting the code to use A/B instead (2.x occasionally uses A/B) but got comments about it and reverted.

Multicompare:
In 2.x compares are "not supported" if more than two revs are selected. 
Implemented in #4564, there are likely some other issues/PRs involved too.

I favor multi-compare over base compare but do not want them to be exclusive. If you meant to only show BASE for two revisions, that is fine. (The last two selected I assume.) Just keep the presentation clear.

**Comment by gerhardol** [comment]
Added #7561 that partially implements this suggestion

@mdonatas 
I took a look at your repo and see that you are maintaining your changes still, nice.
You are welcome to submit your PR, I just added what was simple and what be sufficient for me in most situations.

**Comment by gerhardol** [comment]
@mstv @mdonatas 
I pushed an alternative implementation of BASE-First(A) diff, using an artificial commit

Pro: Less intrusive, affects basically only RevisionDiff and FileStatusList
Con: Another additional artificial diff to handle

feature/i4154-base-a-artificial 3375f47a1bfa008aa65db2fe1898ee775ad805eb

![image](https://user-images.githubusercontent.com/6248932/71635059-6a77cb00-2c21-11ea-8f36-93914847cc78.png)

**Comment by mstv** [comment]
This looks promising.

**Comment by mstv** [comment]
These additional diffs are very useful. Though they can cause a little confusion if there are no differences between the selected revisions. (I use this quite often in order check the correctness of a merge or a rebase.)
Then the section of the empty main diff is not displayed:

![grafik](https://user-images.githubusercontent.com/36601201/72171273-f1bc0000-33d2-11ea-96d7-580d743fe5c6.png)

i.e. the highlighted section below was omitted above because there was no diff matching the filter:

![grafik](https://user-images.githubusercontent.com/36601201/72171388-2a5bd980-33d3-11ea-8736-80e7c20da275.png)

**Comment by gerhardol** [comment]
>i.e. the highlighted section below was omitted above because there was no diff matching the filter:

The section would then seem to not show contents when expanding. But I also believe that this is better. (Should be a general change, not just in this situation.)

**Comment by mstv** [comment]
In addition, I'd use `"Diff branch-a @abcd with branch-b @1234"` for all three sections and remove the trailing colon, i.e. not `"... @1234:"` anymore.
For multi-selections, it often is not ovbious which revisions are used as `a` and `b`.

**Comment by pmiossec** [comment]
> For multi-selections, it often is not ovbious which revisions are used as a and b.

I was thinking about that yesterday and I don't know if it's possible but I imagined that it would be more understandable if we could change the line selection color of the first selected commit to the diff color of 'a' (sort of red) and the second to the diff color of 'b' (sort of green).

**Comment by pmiossec** [comment]
Or perhaps display on the fly two references 'diff/a' and 'diff/b' on the 2 selected commits.

**Comment by gerhardol** [comment]
> For multi-selections, it often is not ovbious which revisions are used as a and b

The GE UI tries to use _Selected_ for last selected, for diffs we use _First_ as an anology. A/B is not really described in Git documentation other than as examples. I have tried discussing this with users and advanced users prefer A/B, less knowledgeable has no idea what is what. Especially Selected is used in many

So if it is changed, it should be consistently _First (A)_ and _Selected (B)_. _(A)_ and _(B)_ could be added to the revision names. Diff colors will be harder, but the control should maybe be updated to allow minimize by default, no calculation  by default etc (right now it can be collapsed just by double clicks).

**Comment by RussKie** [comment]
> Though they can cause a little confusion if there are no differences between the selected revisions.

This can be fixed by leveraging features of OS ListView control which are missing in Winforms implementation, and which are coming in WinForms .NET.
Namely use "subtitles" (https://github.com/dotnet/winforms/issues/2655) or "footers" (https://github.com/dotnet/winforms/issues/2653).

We will need to write few p/invokes for that. I may be able to look into this, if you want.

**Comment by RussKie** [comment]
> I was thinking about that yesterday and I don't know if it's possible but I imagined that it would be more understandable if we could change the line selection color of the first selected commit to the diff color of 'a' (sort of red) and the second to the diff color of 'b' (sort of green).

Could you do an image mock showing this? I don't quite understand what you have in mind

**Comment by RussKie** [comment]
> ![grafik](https://user-images.githubusercontent.com/36601201/72171388-2a5bd980-33d3-11ea-8736-80e7c20da275.png)

What are these "@5963"?

**Comment by mstv** [comment]
> What are these "@5963"?

They are very shortened commit hashs.

**Comment by RussKie** [comment]
We should be consistent and use the central hash-shortening strategy, which is via `ObjectId.ToShortString()`

**Comment by gerhardol** [comment]
> We should be consistent and use the central hash-shortening strategy, which is via ObjectId.ToShortString()

It is using ObjectId.ToShortString(4)
RevisionDiff is the only place shortening to less than 8 (in the group name and reset/diff menus, see #7626).

In this case, the sha is used to identify manually, so I believe it is better use of screen estate.

**Comment by RussKie** [comment]
> > Though they can cause a little confusion if there are no differences between the selected revisions.
> 
> This can be fixed by leveraging features of OS ListView control which are missing in Winforms implementation, and which are coming in WinForms .NET.
> Namely use "subtitles" ([dotnet/winforms#2655](https://github.com/dotnet/winforms/issues/2655)) or "footers" ([dotnet/winforms#2653](https://github.com/dotnet/winforms/issues/2653)).
> 
> We will need to write few p/invokes for that. I may be able to look into this, if you want.

Spend an evening and I have a POC that displays subtitle
```diff
diff --git a/GitUI/UserControls/FileStatusList.cs b/GitUI/UserControls/FileStatusList.cs
index b7732398e..d39fd0483 100644
--- a/GitUI/UserControls/FileStatusList.cs
+++ b/GitUI/UserControls/FileStatusList.cs
@@ -894,6 +894,7 @@ private void UpdateFileStatusListView(bool updateCausedByFilter = false)
                     FileStatusListView.Groups.Add(group);
+                    FileStatusListView.SeGrouptSubtitle(group, "bla");
                 }
```

![image](https://user-images.githubusercontent.com/4403806/73607042-6286be80-4605-11ea-8a96-72312a3d2440.png)


Please let me know if you have a usecase, and I'll create a proper PR or add it to an existing PR.

**Comment by gerhardol** [comment]
I believe I prefer one line only, but with tooltip
Also possibility to collapse on default (follow up: calculate diff when expanding)

For the existing BASE diff, I probably want to add that though in the existing line
And better tooltip in settings

For BASE->First, quite a lot in especially RevisionDiff must be adopted. I want to change the API slightly different to mdonatas
BASE to selected compare is of great use, that is 95% of what is useful (at least for me).

**Comment by gerhardol** [comment]
Status update:
I am to create a PR, still some issues to handle multiple selected commits in the FileStatusList (just the menus remain, there are a lot of changes needed).
The presentation separates the common files and those that are different.

Example for the following commit (with filenames explaining what is changed in the branches)

![image](https://user-images.githubusercontent.com/6248932/77249473-83e94b80-6c41-11ea-886c-81fde3a76d9c.png)

![image](https://user-images.githubusercontent.com/6248932/77249501-b8f59e00-6c41-11ea-8b62-8bc9e55f71d4.png)

It is the changes to "first", branch x in this case that requires changes. There are not so many places that must be changed, but a lot that should be changed. If dropping the BASE->A diff, the need for "should" minimizes and maybe the presentation is clearer. I want it though, but what do others think? Especially you @mdonatas and @mstv.
(Will push a branch soon based on other PRs soon, maybe open a PR.) Then we can discuss what data structures to really use too.

**Comment by mstv** [comment]
Looks useful.

You are right, the `BASE -> A` diff is not really necessary -- especially if there was a possibility to `Swap A/B` for a selection of two commits.

**Comment by gerhardol** [comment]
> You are right, the BASE -> A diff is not really necessary 

Not required, but nice _if you understand the difflist_. I do, but is it confusing for other users?
(I can live with the hours wasted to revert the changes prepared, but handling a object with (GitItemStatus, firstRevision, selectedRevision) has its merits too).
 
> especially if there was a possibility to Swap A/B for a selection of two commits.

Also seeing a need for that occasionally. Of course, separate.

**Comment by gerhardol** [comment]
I have pushed my implementation for this:

https://github.com/gerhardol/gitextensions/tree/feature/i4154-revdiff-base-a

This also have some other submitted PRs prior to the actual changes, as well as some cleanups after. To be separated later (I will likely update this branch some more before submitting a PR).

The major internal change is that instead of saving GitItemStatus per file, the first(A)/parent per group (mostly) and Revision separately, they are all combined in GitItemStatusWithParent. This object can be passed around. This is something I would like to use in other situations like the FileViewer, to unify and simplify handling of line patches for instance. (This PR depends on #7825 and the followup to #7825 depends on the not yet submitted PR).

The changes related to this enhancement has been split in several commits.
_RevDiff: Show common files for BASE->A_ is The change that changes the internal structure and adds the unique/common diff. The following changes the data structure to larger and smaller extent. The BASE->A common-diff could be excluded also if the other are included, if that is hard for the user to understand.

A comment is that I would like to collapse other than the first(A)->selected(B) diff. It has happened that I have been confused myself (for several seconds at least) when looking at the diffs.
However, collapsing by default (and avoid git-diff unless expanded) is not supported in FileStatusList.
Also, empty groups are always hidden.

**Comment by mstv** [comment]
Very clear, sound and comprehensive!
I think with the following group headings it will be user-understandable, especially if we manage to collapse the additional groups. (Although there is no strong need anymore because of the keywords "Unique" and "Common".)
A workaround for hidden empty group could be a placeholder file _"\<identical\>"_.

E.g. let be selected @11111111 "first sel" and @22222222 "last sel", common base @00000000 "ancestor". Then use `a/` and `b/` to match the preamble in the FileDiff:
```
Diff with a/11111111: first sel
   ...
Unique diff BASE to b/22222222: last sel
Unique diff BASE to b/11111111: first sel
Common diff with BASE a/00000000: ancestor
```

**Comment by gerhardol** [comment]
Note that a and b in the commits in the diff command are the first/second commit, so for instance for "BASE->A" will the difftool menu have a=BASE and b=A.

Another issue is that empty groups are hidden. Maybe changing that is something that really should be done, also limits the confusion if groups are hidden. For instance if two commits are identical, just the Common diff group is shown so you have to read the group name. 

I will change the names to something like the following

```
Diff with a/11111111: first sel
   ...
Unique diff BASE with b/22222222: last sel
Unique diff BASE with a/11111111: first sel
Common diff with BASE/00000000: ancestor
```

**Comment by mstv** [comment]
I understood why you have chosen the group names. But it will confuse users IMO: `a/` and `b/` must not have a different meaning in the FileStatusList than in the FileDiff. (`A` / `B` would not be better).
Not such critical: "diff BASE with ... sel" is not exact: The diff has the opposite direction.

**Comment by gerhardol** [comment]
> But it will confuse users IMO: a/ and b/ must not have a different meaning in the FileStatusList than in the FileDiff. 

This could be changed in diff/reset dropdown (will have to add the description in GitItemStatusWithParents, a little weird though), but I would be hesitant to change in the patch shown in FileViewer, so the mismatch will just be somewhere else. So will not change that now.

For preposition use, there were a discussion recently. I tried to use "diff(erence) to" to indicate the direction, but that failed on the tiny detail that it was incorrect English usage. "difference from" is OK though.

![image](https://user-images.githubusercontent.com/6248932/77416251-f6823480-6dc3-11ea-9297-75e0012abcd8.png)

Updated proposal:
```
Diff with a/11111111: first sel
   ...
Unique diff BASE with b/22222222: last sel
Unique diff BASE with a/11111111: first sel
Common diff from BASE/00000000: ancestor
```

**Comment by mstv** [comment]
Your screenshot is the best example why all `unique dfif BASE with` should mention `b/`:
`Unique diff BASE with b/1aa252f6: feature/localize-error`
`BASE` itselfs is `a/` always.
The menu items have already matched the patch shown.

```
Diff with a/11111111: first sel
   ...
Unique diff BASE with b/22222222: last sel
Unique diff BASE with b/11111111: first sel
Common diff with BASE a/00000000: ancestor
```
